### PR TITLE
Bump Kani version to 0.51.0

### PR DESCRIPTION
**Pull request by karkhaz**
For reference, here is the auto-generated changelog

## What's Changed
* Upgrade toolchain to 2024-04-18 and improve toolchain workflow by @celinval in https://github.com/model-checking/kani/pull/3149
* Automatic toolchain upgrade to nightly-2024-04-19 by @github-actions in https://github.com/model-checking/kani/pull/3150
* Stabilize cover statement and update contracts RFC by @celinval in https://github.com/model-checking/kani/pull/3091
* Automatic toolchain upgrade to nightly-2024-04-20 by @github-actions in https://github.com/model-checking/kani/pull/3154
* Bump tests/perf/s2n-quic from `2d5e891` to `5f88e54` by @dependabot in https://github.com/model-checking/kani/pull/3140
* Automatic cargo update to 2024-04-22 by @github-actions in https://github.com/model-checking/kani/pull/3157
* Automatic toolchain upgrade to nightly-2024-04-21 by @github-actions in https://github.com/model-checking/kani/pull/3158
* Bump tests/perf/s2n-quic from `5f88e54` to `9730578` by @dependabot in https://github.com/model-checking/kani/pull/3159
* Fix cargo audit error by @jaisnan in https://github.com/model-checking/kani/pull/3160
* Fix cbmc-update CI job by @tautschnig in https://github.com/model-checking/kani/pull/3156
* Automatic cargo update to 2024-04-29 by @github-actions in https://github.com/model-checking/kani/pull/3165
* Bump tests/perf/s2n-quic from `9730578` to `1436af7` by @dependabot in https://github.com/model-checking/kani/pull/3166
* Do not assume that ZST-typed symbols refer to unique objects by @tautschnig in https://github.com/model-checking/kani/pull/3134
* Fix copyright check for `expected` tests by @adpaco-aws in https://github.com/model-checking/kani/pull/3170
* Remove kani::Arbitrary from the modifies contract instrumentation by @feliperodri in https://github.com/model-checking/kani/pull/3169
* Automatic cargo update to 2024-05-06 by @github-actions in https://github.com/model-checking/kani/pull/3172
* Bump tests/perf/s2n-quic from `1436af7` to `6dd41e0` by @dependabot in https://github.com/model-checking/kani/pull/3174
* Avoid unnecessary uses of Location::none() by @tautschnig in https://github.com/model-checking/kani/pull/3173

**Full Changelog**: https://github.com/model-checking/kani/compare/kani-0.50.0...kani-0.51.0

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,20 @@ This file contains notable changes (e.g. breaking changes, major changes, etc.)
...
 
+## [0.51.0]
+
+## What's Changed
+
+* Do not assume that ZST-typed symbols refer to unique objects by @tautschnig in https://github.com/model-checking/kani/pull/3134
+* Remove kani::Arbitrary from the modifies contract instrumentation by @feliperodri in https://github.com/model-checking/kani/pull/3169
+* Emit source locations whenever possible to ease debugging and coverage reporting by @tautschnig in https://github.com/model-checking/kani/pull/3173
+* Fix cargo audit error by @jaisnan in https://github.com/model-checking/kani/pull/3160
+* Rust toolchain is upgraded to nightly-2024-04-21 by @celinval
```
Comment (adpaco-aws) [inline CHANGELOG.md:15]:
```suggestion
* Rust toolchain upgraded to `nightly-2024-04-21` by @celinval
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,20 @@ This file contains notable changes (e.g. breaking changes, major changes, etc.)
...
 This file was introduced starting Kani 0.23.0, so it only contains changes from version 0.23.0 onwards.
 
+## [0.51.0]
+
+## What's Changed
+
+* Do not assume that ZST-typed symbols refer to unique objects by @tautschnig in https://github.com/model-checking/kani/pull/3134
+* Remove kani::Arbitrary from the modifies contract instrumentation by @feliperodri in https://github.com/model-checking/kani/pull/3169
+* Emit source locations whenever possible to ease debugging and coverage reporting by @tautschnig in https://github.com/model-checking/kani/pull/3173
+* Fix cargo audit error by @jaisnan in https://github.com/model-checking/kani/pull/3160
```
Comment (adpaco-aws) [inline CHANGELOG.md:14]: I'd remove this item because it's a dependency upgrade

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,20 @@ This file contains notable changes (e.g. breaking changes, major changes, etc.)
 
 This file was introduced starting Kani 0.23.0, so it only contains changes from version 0.23.0 onwards.
 
+## [0.51.0]
+
+## What's Changed
+
+* Do not assume that ZST-typed symbols refer to unique objects by @tautschnig in https://github.com/model-checking/kani/pull/3134
+* Remove kani::Arbitrary from the modifies contract instrumentation by @feliperodri in https://github.com/model-checking/kani/pull/3169
```
Comment (adpaco-aws) [inline CHANGELOG.md:12]:
```suggestion
* Remove `kani::Arbitrary` from the `modifies` contract instrumentation by @feliperodri in https://github.com/model-checking/kani/pull/3169
```

**Review by adpaco-aws — APPROVED** [review]

**Comment by karkhaz** [comment]
For the record, solver runtime for
`s2n-quic/quic/s2n-quic-core/inet::checksum::tests::differential` has regressed from 97s -> 	173s (+79.085%). I'm merging.

**Pull request merged**